### PR TITLE
[Fix](Join) Fix the bug of outer join function under vectorization

### LINKED ISSUE #7901: [Feature] [Vectorized] Some Join opt in vec exec engine
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

1. shoule adjust the out join column to nullable
    `select * from t1 left join t2 on t1.a = t2.a` 
    all slots of the t2 shoule be nullable change in join node.

2. Adjust the join column of non nullaware join to non nullable
    `select * from t1 join t2 on t1.a = t2.a`  if a is a nullable column shoule change to not nullable, because of join column `=` will skip the null.

3. Useless columns do not need to be output， here need a projection to reduce the column output to speed up
the join operator.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by EmmyMiao87**
# Proposed changes

Closed #7901

## Problem Summary

This pr mainly fixes the following problems:
1. Solve the query combined with inline view and outer join. After adding a tuple to the join operator, the position of the `tupleisnull` function is inconsistent with the row storage. Currently the vectorized `tupleisnull` will be calculated in the HashJoinNode.computeOutputTuple() function.
2. Column nullable property error problem. At present, once the outer join occurs, the column on the null-side side will be planned to be nullable in the semantic parsing stage.

### About 1:
For example：
```
select * from (select a as k1 from test) tmp right join b on tmp.k1=b.k1
```
At this time, the nullable property of column k1 in the `tmp` inline view should be true.

In the vectorized code, the virtual `tableRef` of tmp will be used in constructing the output tuple of HashJoinNode (specifically, the function HashJoinNode.computeOutputTuple()). So the **correctness** of the column nullable property of this tableRef is very important.
In the above case, since the tmp table needs to perform a right join with the b table, as a null-side tmp side, it is necessary to change the column attributes involved in the tmp table to nullable.

In non-vectorized code, since the virtual tableRef tmp is not used at all, it uses the `TupleIsNull` function in `outputsmp` to ensure data correctness.
That is to say, the a column of the original table test is still non-null, and it does not affect the correctness of the result.

### About 2
The vectorized nullable attribute requirements are very strict. 
Outer join will change the nullable attribute of the join column, thereby changing the nullable attribute of the column in the upper operator layer by layer. 
Since FE has no mechanism to modify the nullable attribute in the upper operator tuple layer by layer after the analyzer. 
So at present, we can only preset the attributes before the lower join as nullable in the analyzer stage in advance, so as to avoid the problem. 
(At the same time, be also wrote some evasive code in order to deal with the problem of null to non-null.)

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
4. Has document been added or modified: (Yes/No/No Need)
5. Does it need to update dependencies: (Yes/No)
6. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morningman** [comment]
Could you add some ut for this?

File: gensrc/thrift/PlanNodes.thrift [hunk 1]
```diff
@@ -410,6 +410,10 @@ struct THashJoinNode {
 
   // hash output column
   6: optional list<Types.TSlotId> hash_output_slot_ids
+
+  7: optional list<Exprs.TExpr> srcExprList
```
Comment (HappenLee) [inline gensrc/thrift/PlanNodes.thrift:414]: use `src_expr_list`

File: gensrc/thrift/PlanNodes.thrift [hunk 2]
```diff
@@ -409,6 +409,10 @@ struct THashJoinNode {
...
+
+  8: optional Types.TTupleId voutput_tuple_id
```
Comment (morrySnow) [inline gensrc/thrift/PlanNodes.thrift:415]: move it to TPlanNode

**Comment by morningman** [comment]
Fixed in #10323

**Review by morningman — APPROVED** [review]
LGTM

**Review by morningman — CHANGES_REQUESTED** [review]
hold for a second, there is still some problem

**Comment by EmmyMiao87** [comment]
The contents of this pr are merged into the new pr#10437  and submitted together

**Pull request closed without merging**